File: Semestru 4/MPP/lab/java/lab3/Persistence/src/main/java/org/example/repository/db/SignupDbRepository.java
package org.example.repository.db;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.example.domain.Child;
import org.example.domain.Event;
import org.example.domain.Signup;
import org.example.domain.Tuple;
import org.example.repository.interf.ISignupRepository;
import org.example.repository.utils.CommonUtils;


import java.sql.Connection;
import java.sql.SQLException;
import java.util.ArrayList;
import java.util.List;
import java.util.Properties;

public class SignupDbRepository implements ISignupRepository {
    private CommonUtils Utils;
    private static final Logger logger = LogManager.getLogger();

    public SignupDbRepository(Properties properties) {
        logger.info("Initializing SignupDbRepository with properties: {}", properties);
        this.Utils = new CommonUtils(properties);
    }
    @Override
    public Iterable<Signup> GetAllByChildId(Integer id) {
        logger.traceEntry("Looking for signup with child id: {} ", id);
        Connection connection = this.Utils.GetConnection();
        List<Signup> signups = new ArrayList<>();
        try (var statement = connection.prepareStatement("select * from signups where id_child = ? "))
        {
            statement.setInt(1, id);
            try (var resultSet = statement.executeQuery()) {
                if (resultSet.next()) {
                    signups.add( (Signup) new Signup()
                            .SetId(new Tuple<>(resultSet.getInt(1), resultSet.getInt(2)))
                    );
                }
            }
        } catch (SQLException e) {
            logger.error(e);
            System.out.println("Error" + e);
        }
        logger.traceExit();
        return signups;
    }

    @Override
    public Iterable<Signup> GetAllByEventId(Integer id) {
        logger.traceEntry("Looking for signup with event id: {} ", id);
        Connection connection = this.Utils.GetConnection();
        List<Signup> signups = new ArrayList<>();
        try (var statement = connection.prepareStatement("select * from signups where id_event = ? "))
        {
            statement.setInt(1, id);
            try (var resultSet = statement.executeQuery()) {
                if (resultSet.next()) {
                    signups.add( (Signup) new Signup()
                            .SetId(new Tuple<>(resultSet.getInt(1), resultSet.getInt(2)))
                    );
                }
            }
        } catch (SQLException e) {
            logger.error(e);
            System.out.println("Error" + e);
        }
        logger.traceExit();
        return signups;
    }

    @Override
    public Signup FindById(Tuple<Integer, Integer> integerIntegerTuple) {
        logger.traceEntry("Looking for signup with id: {} {}", integerIntegerTuple.GetFirst(), integerIntegerTuple.GetSecond());
        Connection connection = this.Utils.GetConnection();
        Signup signup = null;
        try (var statement = connection.prepareStatement(
                "SELECT c.id AS child_id, c.name AS child_name, c.cnp AS child_cnp, " +
                "e.id AS event_id, e.name AS event_name, e.min_age, e.max_age " +
                "FROM signups s " +
                "JOIN children c ON s.id_child = c.id " +
                "JOIN events e ON s.id_event = e.id "))
        {
            statement.setInt(1, integerIntegerTuple.GetFirst());
            statement.setInt(2, integerIntegerTuple.GetSecond());
            try (var resultSet = statement.executeQuery()) {
                if (resultSet.next()) {
                   signup = (Signup) new Signup(
                           (Child) new Child(
                                   resultSet.getString(2),
                                   resultSet.getString(3)
                           ).SetId(resultSet.getInt(1)),
                           (Event) new Event(
                                   resultSet.getString(5),
                                   resultSet.getInt(6),
                                   resultSet.getInt(7)
                           ).SetId(resultSet.getInt(4))
                   ).SetId(new Tuple(resultSet.getInt(1), resultSet.getInt(4)));
                }
            }
        } catch (SQLException e) {
            logger.error(e);
            System.out.println("Error" + e);
        }
        logger.traceExit();
        return signup;
    }

    @Override
    public Iterable<Signup> FindAll() {
        logger.traceEntry("Getting all of the signups.");
        Connection connection = this.Utils.GetConnection();
        List<Signup> signups = new ArrayList<>();
        try (var statement = connection.prepareStatement(
                "SELECT c.id AS child_id, c.name AS child_name, c.cnp AS child_cnp, " +
                        "e.id AS event_id, e.name AS event_name, e.min_age, e.max_age " +
                        "FROM signups s " +
                        "JOIN children c ON s.id_child = c.id " +
                        "JOIN events e ON s.id_event = e.id ")) {
            try (var resultSet = statement.executeQuery()) {
                while (resultSet.next()) {
                    logger.debug("Adding signup for " + resultSet.getString(2) + " " + resultSet.getString(5));
                    signups.add(
                            (Signup) new Signup(
                                    (Child) new Child(
                                            resultSet.getString(2),
                                            resultSet.getString(3)
                                    ).SetId(resultSet.getInt(1)),
                                    (Event) new Event(
                                            resultSet.getString(5),
                                            resultSet.getInt(6),
                                            resultSet.getInt(7)
                                    ).SetId(resultSet.getInt(4))
                            ).SetId(new Tuple(resultSet.getInt(1), resultSet.getInt(4)))
                    );
                }
            }
        } catch (SQLException e) {
            logger.error(e);
            System.out.println("Error" + e);
        }
        logger.traceExit();
        return signups;
    }

    @Override
    public Signup Save(Signup entity) {
        logger.traceEntry("Saving signup " + entity);
        Connection connection = this.Utils.GetConnection();
        try (var statement = connection.prepareStatement("insert into signups (id_child, id_event) values (?, ?)"))
        {
            statement.setInt(1, entity.GetChild().GetId());
            statement.setInt(2, entity.GetEvent().GetId());
            int result = statement.executeUpdate();
            return (Signup) entity.SetId(new Tuple<Integer, Integer>(entity.GetChild().GetId(), entity.GetEvent().GetId()));
        } catch (SQLException e) {
            logger.error(e);
            System.err.println("Error " + e);
        }
        logger.traceExit();
        return null;
    }

    @Override
    public Signup Update(Signup entity) {
        return null;
    }

    @Override
    public Signup Delete(Tuple<Integer, Integer> integerIntegerTuple) {
        return null;
    }
}

File: Semestru 4/MPP/lab/.net/lab3/lab3/Persistence/repo/implementation/SignupDbRepository.cs
using Model.app.domain;
using Persistence.app.repo.@interface;
using Persistence.app.utils;
using log4net;
using System.Data.SQLite;

namespace Persistence.app.repo.implementation
{
	public class SignupDbRepository : ISignupRepository
	{
        private CommonUtils Utils;
		private static readonly ILog _logger = LogManager.GetLogger(typeof(EventDbRepository));

		public SignupDbRepository()
		{
			_logger.Info("Initialized SignupDbRepository.");
			this.Utils = new CommonUtils();
		}

		public Signup Create(Signup e)
		{
			_logger.Info($"Saving the signup: {e.Child.Name}, {e.Event.Name}");
			var connection = this.Utils.GetConnection();
			try
			{
				_logger.Info("Succesfully connected to the db.");
				if (connection.State != System.Data.ConnectionState.Open)
					connection.Open();
				using (var statement = new SQLiteCommand("insert into signups (id_child, id_event) values (@id_child, @id_event)", connection))
				{
					statement.Parameters.Clear();
					statement.Parameters.AddWithValue("@id_child", e.Child.Id);
					statement.Parameters.AddWithValue("@id_event", e.Event.Id);
					int rows = statement.ExecuteNonQuery();
					_logger.Info($"Affected rows: {rows}");
					return (Signup) e.SetId(Tuple.Create(e.Child.Id, e.Event.Id));
				}
			}
			catch (Exception ex)
			{
				_logger.Info($"Caught exception: " + ex);
				Console.WriteLine(ex.Message);
			}
			return null!;
		}

		public Signup? Delete(Tuple<int, int> id)
		{
			throw new NotImplementedException();
		}

		public IEnumerable<Signup> GetAll()
		{
			_logger.Info($"Searching for all signups ");
			var connection = this.Utils.GetConnection();
			try
			{
				_logger.Info("Succesfully connected to the db.");
				if (connection.State != System.Data.ConnectionState.Open)
					connection.Open();
				List<Signup> signups = new List<Signup>();
				using (var query = new SQLiteCommand("select * from signups", connection))
				{
					using (var reader = query.ExecuteReader())
					{
						while (reader.Read())
							signups.Add(
								(Signup)new Signup(
									
								).SetId(Tuple.Create(
									reader.GetInt32(reader.GetOrdinal("id_child")),
									reader.GetInt32(reader.GetOrdinal("id_event"))
								))
						   );
						return signups;
					}
				}
			}
			catch (Exception ex)
			{
				_logger.Info($"Caught exception: " + ex);
				Console.WriteLine(ex.Message);
			}
			return null!;
		}

		public IEnumerable<Signup> GetAllByChildId(int id)
		{
			_logger.Info($"Searching for all signups ");
			var connection = this.Utils.GetConnection();
			try
			{
				_logger.Info("Succesfully connected to the db.");
				if (connection.State != System.Data.ConnectionState.Open)
					connection.Open();
				List<Signup> signups = new List<Signup>();
				using (var query = new SQLiteCommand(
						"SELECT c.id AS child_id, c.name AS child_name, c.cnp AS child_cnp, " +
						"e.id AS event_id, e.name AS event_name, e.min_age, e.max_age " +
						"FROM signups s " +
						"JOIN children c ON s.id_child = c.id " +
						"JOIN events e ON s.id_event = e.id " +
						"WHERE c.id = @id_child", connection))
				{
					query.Parameters.AddWithValue("@id_child", id);
					using (var reader = query.ExecuteReader())
					{
						while (reader.Read())
						{
							var child = new Child
							{
								Id = reader.GetInt32(reader.GetOrdinal("child_id")),
								Name = reader.GetString(reader.GetOrdinal("child_name")),
								Cnp = reader.GetString(reader.GetOrdinal("child_cnp"))
							};

							var event_ = new Event
							{
								Id = reader.GetInt32(reader.GetOrdinal("event_id")),
								Name = reader.GetString(reader.GetOrdinal("event_name")),
								MinAge = reader.GetInt32(reader.GetOrdinal("min_age")),
								MaxAge = reader.GetInt32(reader.GetOrdinal("max_age"))
							};
							signups.Add((Signup)new Signup(child, event_).SetId(Tuple.Create(child.Id, event_.Id)));
						}
						return signups;
					}
				}
			}
			catch (Exception ex)
			{
				_logger.Info($"Caught exception: " + ex);
				Console.WriteLine(ex.Message);
			}
			return null!;
		}

		public IEnumerable<Signup> GetAllByEventId(int id)
		{
			_logger.Info($"Searching for all signups ");
			var connection = this.Utils.GetConnection();
			try
			{
				_logger.Info("Succesfully connected to the db.");
				if (connection.State != System.Data.ConnectionState.Open)
					connection.Open();
				List<Signup> signups = new List<Signup>();
				using (var query = new SQLiteCommand(
						"SELECT c.id AS child_id, c.name AS child_name, c.cnp AS child_cnp, " +
						"e.id AS event_id, e.name AS event_name, e.min_age, e.max_age " +
						"FROM signups s " +
						"JOIN children c ON s.id_child = c.id " +
						"JOIN events e ON s.id_event = e.id " +
						"WHERE e.id = @id_event", connection))
				{
					query.Parameters.AddWithValue("@id_event", id);
					using (var reader = query.ExecuteReader())
					{
						while (reader.Read())
						{
							var child = new Child
							{
								Id = reader.GetInt32(reader.GetOrdinal("child_id")),
								Name = reader.GetString(reader.GetOrdinal("child_name")),
								Cnp = reader.GetString(reader.GetOrdinal("child_cnp"))
							};

							var event_ = new Event
							{
								Id = reader.GetInt32(reader.GetOrdinal("event_id")),
								Name = reader.GetString(reader.GetOrdinal("event_name")),
								MinAge = reader.GetInt32(reader.GetOrdinal("min_age")),
								MaxAge = reader.GetInt32(reader.GetOrdinal("max_age"))
							};
							signups.Add((Signup)new Signup(child, event_).SetId(Tuple.Create(child.Id, event_.Id)));
						}
						return signups;
					}
				}
			}
			catch (Exception ex)
			{
				_logger.Info($"Caught exception: " + ex);
				Console.WriteLine(ex.Message);
			}
			return null!;
		}

		public Signup? GetById(Tuple<int, int> id)
		{
			_logger.Info($"Searching for all signups ");
			var connection = this.Utils.GetConnection();
			try
			{
				_logger.Info("Succesfully connected to the db.");
				if (connection.State != System.Data.ConnectionState.Open)
					connection.Open();
				List<Signup> signups = new List<Signup>();
				using (var query = new SQLiteCommand("select * from signups where id_child = @id_child and id_event = @id_event", connection))
				{
					query.Parameters.AddWithValue("@id_child", id.Item1);
					query.Parameters.AddWithValue("@id_event", id.Item2);
					using (var reader = query.ExecuteReader())
					{
						if (reader.Read())
							return (Signup)new Signup(

								).SetId(Tuple.Create(
									reader.GetInt32(reader.GetOrdinal("id_child")),
									reader.GetInt32(reader.GetOrdinal("id_event"))
								)
						   );
					}
				}
			}
			catch (Exception ex)
			{
				_logger.Info($"Caught exception: " + ex);
				Console.WriteLine(ex.Message);
			}
			return null!;
		}

		public Signup? Update(Signup e)
		{
			throw new NotImplementedException();
		}

		public IEnumerable<Signup> GetAllJoin()
		{
			_logger.Info($"Searching for all signups joined ");
			var connection = this.Utils.GetConnection();
			try
			{
				_logger.Info("Succesfully connected to the db.");
				if (connection.State != System.Data.ConnectionState.Open)
					connection.Open();
				List<Signup> result = new List<Signup>();
				using (var query = new SQLiteCommand(
					"SELECT c.id AS child_id, c.name AS child_name, c.cnp AS child_cnp, " +
					"e.id AS event_id, e.name AS event_name, e.min_age, e.max_age " +
					"FROM signups s " +
					"JOIN children c ON s.id_child = c.id " +
					"JOIN events e ON s.id_event = e.id", connection))
				{
					using (var reader = query.ExecuteReader())
					{
						while (reader.Read())
						{
							var child = new Child
							{
								Id = reader.GetInt32(reader.GetOrdinal("child_id")),
								Name = reader.GetString(reader.GetOrdinal("child_name")),
								Cnp = reader.GetString(reader.GetOrdinal("child_cnp"))
							};

							var event_ = new Event
							{
								Id = reader.GetInt32(reader.GetOrdinal("event_id")),
								Name = reader.GetString(reader.GetOrdinal("event_name")),
								MinAge = reader.GetInt32(reader.GetOrdinal("min_age")),
								MaxAge = reader.GetInt32(reader.GetOrdinal("max_age"))
							};
							result.Add((Signup)new Signup(child, event_).SetId(Tuple.Create(child.Id, event_.Id)));
						}
						return result;
					}
				}
			}
			catch (Exception ex)
			{
				_logger.Info($"Caught exception: " + ex);
				Console.WriteLine(ex.Message);
			}
			return null!;
		}

	}
}
